src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+                        break;
+                    default:
+                        break;
+                }
+            }
+        } catch (MessagePackException | IOException ex) {
+            throw new RuntimeException("Error reading MessagePack data.", ex);
+        }
+
+        return hubMessages.toArray(new HubMessage[hubMessages.size()]);
+    }
+    
+    @Override
+    public String writeMessage(HubMessage hubMessage) {
+        HubMessageType messageType = hubMessage.getMessageType();
+        
+        try {
+            switch (messageType) {
+                case INVOCATION:
+                    return writeInvocationMessage((InvocationMessage) hubMessage);
+                case STREAM_ITEM:
+                    return writeStreamItemMessage((StreamItem) hubMessage);
+                case COMPLETION:
+                    return writeCompletionMessage((CompletionMessage) hubMessage);
+                case STREAM_INVOCATION:
+                    return writeStreamInvocationMessage((StreamInvocationMessage) hubMessage);
+                case CANCEL_INVOCATION:
+                    return writeCancelInvocationMessage((CancelInvocationMessage) hubMessage);
+                case PING:
+                    return writePingMessage((PingMessage) hubMessage);
+                case CLOSE:
+                    return writeCloseMessage((CloseMessage) hubMessage);
+                default:
+                    throw new RuntimeException(String.format("Unexpected message type: %d", messageType.value));
+            }
+        } catch (MessagePackException | IOException ex) {
+            throw new RuntimeException("Error writing MessagePack data.", ex);
+        }
+    }
+        
+    private HubMessage createInvocationMessage(MessageUnpacker unpacker, InvocationBinder binder, int itemCount) throws IOException {
+        Map<String, String> headers = readHeaders(unpacker);
+        String invocationId = unpacker.unpackString();
+        
+        // For MsgPack, we represent an empty invocation ID as an empty string,
+        // so we need to normalize that to "null", which is what indicates a non-blocking invocation.
+        if (invocationId == null || invocationId.isEmpty()) {
+            invocationId = null;
+        }
+            
+        String target = unpacker.unpackString();
+        
+        Object[] arguments = null;
+        try {
+            List<Class<?>> types = binder.getParameterTypes(target);
+            arguments = bindArguments(unpacker, types);
+        } catch (Exception ex) {
+            return new InvocationBindingFailureMessage(invocationId, target, ex);
+        }
+        
+        // NOTE - C# client accounts for possibility of streamId array being absent, while spec claims it's required - which?
+        Collection<String> streams = readStreamIds(unpacker);
+        
+        // NOTE - C# client applies headers by assigning Message.Headers directly, rather than adding new constructor
```

Comment:
I added a constructor to InvocationMessage (and other HubMessage types) to take Headers, whereas the C# client assigns directly to the field: https://github.com/dotnet/aspnetcore/blob/0d156487aa27df65d68fc7fa31e7b6cb9f4d7581/src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocolWorker.cs#L267-L275

Reply:
Is there a property initializer syntax in Java. If not, I think adding/changing the constructor to a package private class is fine.

Reply:
You can set `object.field = x` for public properties, but as per the other comment I think I'll just simplify the Messages to use one constructor